File: 0046-permutations/0046-permutations.go
func swap(nums []int, i int, j int){
    nums[i],nums[j] = nums[j],nums[i]
}

func rec(i int, n int, answer *[][]int, nums []int){
    if i == n {
        curr := make([]int, len(nums))
        copy(curr,nums)
        *answer = append(*answer, curr)
        return
    }
    for j:=i; j<n; j++ {
        swap(nums,i,j)
        rec(i+1,n,answer,nums)
        swap(nums,i,j)
    }
}

func permute(nums []int) [][]int {
    var answer [][]int
    rec(0,len(nums),&answer,nums)
    return answer
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        size_t n = nums.size();
        vector<vector<int>> answer;
        function<void(int)> rec = [&](int i){
            if(i == n){ 
                answer.push_back(nums);
                return;
            }
            for(int j=i; j<n; ++j){
                swap(nums[i],nums[j]);
                rec(i+1);
                swap(nums[i],nums[j]);
            }
        };
        rec(0);
        return answer;
    }
};